plutus-pab: Adds Swagger support for web-server API

kelizarov opened the pull request:
This change adds a support for swagger using servant-openapi that generates an OpenAPI spec for PAB API with descriptions to provide a non-technical information for how endpoints behave and what they expect in requests and responses.

The swagger-ui is accessible through `swagger/swagger-ui/index.html`
The example of UI:
![telegram-cloud-photo-size-2-5269758382535980541-y](https://user-images.githubusercontent.com/7036666/130748070-674d4b4d-fc4b-4feb-92fb-b90feb7593de.jpg)

Pre-submit checklist:
- Branch
    - [ ] Tests are provided (if possible)
    - [x] Commit sequence broadly makes sense
    - [x] Key commits have useful messages
    - [ ] Relevant tickets are mentioned in commit messages
    - [x] Formatting, materialized Nix files, PNG optimization, etc. are updated
- PR
    - [x] Self-reviewed the diff
    - [x] Useful pull request description
    - [x] Reviewer requested

j-mueller commented on notes/model/UTxO.hsproj/Ledger.hs:
```diff
@@ -36,7 +39,8 @@ data Tx
     }
-    deriving (Show)
+    deriving (Show, Generic)
+    deriving (OpenApi.ToSchema)
```
This code isn't used in the PAB so it doesn't have to be updated

j-mueller reviewed (commented):
Making the API self-describing is generally very useful and will make it much easier to generate clients in other languages. We looked at swagger a while ago and the issue that prevented us from using it was lack of support for sum types. But it looks like a new major version of swagger/open API is out so maybe it's time to revisit this.

How does OpenAPI3 handle sum types? For example what does the schema for `UniswapContracts` look like?

kelizarov commented:
Hey @j-mueller the `servant-openapi` adds support for Sum types by using `oneOf` construction. I've fired up Uniswap PAB instance with these changes, here's how the model looks like in swagger-ui:
![image](https://user-images.githubusercontent.com/7036666/131319296-2eebddcc-1ac5-4b21-bf79-e76e5ed7a031.png)

kelizarov commented:
There's also an issue with representing tuple types in swagger. For example for `txRedeemers` in `Tx` type it represented like this: 
![image](https://user-images.githubusercontent.com/7036666/131471648-201340f5-f846-4a4b-975e-72cd59741fe1.png)

The simple solution could be to declare the `ToSchema` instance manually instead of using the generic deriviation. Let me know if this is critical here.

j-mueller commented:
The `oneOf` construct looks nice, so I'm happy to merge this now.

> issue with representing tuple types in swagger

Sorry what is the issue exactly? That the types of the elements of the list aren't specified? In general, encoding a tuple as a heterogeneous list with a fixed length is fine (that is what a tuple is, and that's the default encoding that `aeson` uses)

michaelpj commented:
This PR adds package dependencies to `plutus-ledger` which are not used, and that's what Hydra complains about: https://github.com/input-output-hk/plutus/pull/3807/files#diff-cb3369701745e982609f4dfe004aa53507b8570ddc9841d1aa4f186109c32725R122-R123

michaelpj commented:
I don't think we should add a dependency on `openapi` to the packages below `plutus-ledger-api`. Those are depended on by the node, and I don't think we want this in the dependency tree for that. You could put them in your orphans module, perhaps.

michaelpj commented on orphans/src/Orphans/OpenApi.hs:
```diff
@@ -0,0 +1,18 @@
+{-# LANGUAGE FlexibleInstances  #-}
+{-# LANGUAGE OverloadedStrings  #-}
+{-# LANGUAGE StandaloneDeriving #-}
+module Orphans.OpenApi () where
```
I'm not sure what the advantage of putting this in a package of its own is. Why not just put it in a module in the package where you use it, or `plutus-ledger`? We already have orphans in there. It also doesn't make much sense to me to have a generic `orphans` package, would we put all orphans for all packages in there?

kelizarov replied:
The idea was to unify the orpans that can be shared with multiple packages. For example it needs for `plutus-pab` and `plutus-contract`.

We could put orphans into `plutus-ledger` but would it make sense to add a dependecy just for orphans alone? Or maybe we can just define the `Orphan` module for each package independetly upon requirement, since now it only is needed for some base types (`aeson` or `freer`)?

michaelpj replied:
Well, if you follow my suggestion about not having `plutus-ledger-api` and `plutus-core` depend on `openapi`, then you'll need a bunch of orphan instances for the types from those packages too. `plutus-ledger` seems like the best place to put all that, since the others depend on it already. Or maybe `plutus-contract`, WDYT @j-mueller ?

j-mueller replied:
Yes I was going to comment on that too - `plutus-ledger` shoud be fine.

michaelpj commented on shell.nix:
```diff
@@ -72,6 +72,7 @@ let
     yq
     z3
     zlib
+    lzma
```
why do we need this?

kelizarov replied:
It's the requirement for `servant-swagger-ui`.

michaelpj replied:
`haskell.nix` should handle native dependencies automatically. What happens if you don't include this?

kelizarov replied:
Checked, it works without specifying it. Removed it.

kelizarov commented:
> This PR adds package dependencies to `plutus-ledger` which are not used, and that's what Hydra complains about: https://github.com/input-output-hk/plutus/pull/3807/files#diff-cb3369701745e982609f4dfe004aa53507b8570ddc9841d1aa4f186109c32725R122-R123

But they're used in these module:
https://github.com/input-output-hk/plutus/pull/3807/files#diff-3792e47803463e7e1b312bd15a54ff370e386fb431ac1d03fb3f68b8bb65c648
https://github.com/input-output-hk/plutus/pull/3807/files#diff-dad89552434f03995ef455b89ec062dd4cc79caf69faf09aeb2ab70d1417cb7b

Is there some other way to include dependencies to packages?

michaelpj commented:
No they aren't. You added `freer-simple`, `semigroups`, and `openapi3`, and you're only using `openapi3`. Which is exactly what the error message said.

kelizarov commented:
> No they aren't. You added `freer-simple`, `semigroups`, and `openapi3`, and you're only using `openapi3`. Which is exactly what the error message said.

My bad, didn't notice it mentioned the other two. Removed them as well.

kelizarov commented:
@j-mueller @michaelpj I've added the changes for orphan instances: put them on `plutus-ledger` package and removed `openapi3` dependency from `plutus-ledger-api` and `plutus-core`.

Fiftyw3bs commented:
I think the merging of this PR has taken enough time. 

I'm stuck at `ToSchema` not accepting Sum types. This PR seems to be the only block standing in my way ;(

j-mueller commented on plutus-pab/src/Plutus/PAB/Webserver/API.hs:
```diff
@@ -30,20 +32,21 @@ type WSAPI =
-    ("healthcheck" :> Get '[JSON] () -- Is the server alive?
-    :<|> ("fullreport" :> Get '[JSON] (FullReport t)) -- Details of the contracts: the signatures and their states.
-    :<|> "contract" :> ("activate" :> ReqBody '[JSON] (ContractActivationArgs t) :> Post '[JSON] ContractInstanceId -- Start a new instance.
+    = "api" :> ("healthcheck" :> Description "Is the server alive?" :> Get '[JSON] ()
```
Nice!

j-mueller reviewed (commented):
I'd like to get a second opinion from @silky before merging, but it looks good to me now.

silky reviewed (changes requested):
Looks great to me! Thanks for doing this work. I've added a couple of small suggestions, then I'm happy to merge!

silky commented on plutus-ledger/src/Ledger/Orphans.hs:
```diff
@@ -45,14 +45,10 @@ instance OpenApi.ToSchema Crypto.XPub where
-
+deriving instance OpenApi.ToSchema (LogMessage JSON.Value)
+deriving instance OpenApi.ToSchema LogLevel
 instance {-# INCOHERENT #-} OpenApi.ToSchema JSON.Value where
```
Sorry @kelizarov - Do you mind removing this last `INCOHERENT` here; then it will be good to merge! :)

kelizarov replied:
Missed that. Thanks for pointing out :)

silky commented:
Hey @kelizarov - All looks great; just one final thing - If you can rebase to master and squish a few of the commits down so the history is a bit cleaner that would be perfect!

michaelpj commented:
@silky you can also just do a squash-merge if you like

silky commented:
Ah thanks @michaelpj ; I looked at 'rebase and merge; only :)
I'll do that!

Fiftyw3bs commented:
Hi @michaelpj 

May I ask how this affects the developers waiting to make use of the solutions in this PR? I, for one, have been stuck for like a week waiting for it to be merged.

michaelpj commented:
It shouldn't affect you. I have generously decided not to revert it and instead to let it be fixed in a subsequent PR :)

paul-outlaw commented:
Kindly incorporate ASAP into next `plutus-starter-devcontainer`: v1.0.10
